File: 041-和为S的两个数字/041-FindNumbersWithSum.py
# -*- coding:utf-8 -*-
class Solution:
    def FindNumbersWithSum(self, array, tsum):
        # write code here
        pool=set()
        val1=float('inf')
        val2=float('inf')
        for val in array:
            diff=tsum-val
            if diff in pool:
                if val1*val2>diff*val:
                    val1,val2=diff,val
            else:
                pool.add(val)
        if val1==float('inf'):
            return []
        return val1,val2

File: 041-和为S的两个数字/041-FindNumbersWithSum.cpp
class Solution {
public:
    vector<int> FindNumbersWithSum(vector<int> array,int sum) {
        vector<int> none;
        if (array.size()==0){
            return none;
        }
        vector<int> ans(2);
        bool tag=true;
        auto begin=array.begin();
        auto end=array.end();
        end--;
        while(begin<end){
            long long csum=(*begin)+(*end);
            if(csum==sum){
                if(tag||(ans[0]*ans[1]>(* begin) * (*end))){
                    ans[0]=*begin;
                    ans[1]=*end;
                    tag=false;
                }
                begin++;
            }
            else if (csum<sum){
                begin++;
            }
            else{
                end--;
            }
        }
        if (tag){
            return none;
        }
        return ans;
    }
};
